File: backend/ProductTracker.Api/Domain/Entities/PurchaseRequest.cs
namespace ProductTracker.Api.Domain.Entities;

public sealed class PurchaseRequest
{
    public Guid Id { get; set; }

    public string RequestNumber { get; set; } = null!;

    public Guid RequestedByUserId { get; set; }
    public User RequestedByUser { get; set; } = null!;

    public DateTime RequestDate { get; set; }

    public PurchaseRequestStatus Status { get; set; } = PurchaseRequestStatus.Draft;

    public string? Description { get; set; }

    public ICollection<PurchaseRequestLine> Lines { get; set; } = new List<PurchaseRequestLine>();

    public Guid? ApprovedByUserId { get; set; }
    public User? ApprovedByUser { get; set; }
    public DateTime? ApprovedAt { get; set; }

    public string? RejectionReason { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? SubmittedAt { get; set; }
}

File: frontend/producttracker-web/src/types/purchaseRequest.ts
export type PurchaseRequestLineResponse = {
    id: string;
    productId: string;
    productName?: string | null;
    quantity: number;
    unitId: string;
    unitCode: string;
    unitName: string;
    requiredDate: string | null;
    notes: string | null;
};

export const PurchaseRequestStatus = {
    Draft: 0,
    Submitted: 1,
    Approved: 2,
    Rejected: 3,
    Cancelled: 4,
} as const;
export type PurchaseRequestStatus =
    (typeof PurchaseRequestStatus)[keyof typeof PurchaseRequestStatus];

export type PurchaseRequestResponse = {
    id: string;
    requestNumber: string;
    requestedByUserId: string;
    requestedByUserName?: string | null;
    requestDate: string;
    status: PurchaseRequestStatus;
    description: string | null;
    createdAt: string;
    submittedAt: string | null;
    approvedByUserId: string | null;
    approvedAt: string | null;
    rejectionReason: string | null;
    lines: PurchaseRequestLineResponse[];
};

export type CreatePurchaseRequestLineRequest = {
    productId: string;
    quantity: number;
    unitId: string;
    requiredDate?: string | null;
    notes?: string | null;
};

export type CreatePurchaseRequestRequest = {
    requestDate: string;
    description?: string | null;
    lines: CreatePurchaseRequestLineRequest[];
};

export type PagedResult<T> = {
    page: number;
    pageSize: number;
    total: number;
    items: T[];
};

export type ListPurchaseRequestsQuery = {
    q?: string;
    status?: PurchaseRequestStatus;
    requestedByUserId?: string;
    fromDate?: string;
    toDate?: string;
    sort?: string;
    page?: number;
    pageSize?: number;
};
